### PR TITLE
ARROW-4622: [C++][Python] MakeDense and MakeSparse in UnionArray should accept a vector of Field

### PR DESCRIPTION
**Pull request by mrkn**
## TODO:

- [x] Write tests for existing behaviors
- [x] Support to supply field names
- [x] union_(field_names, children, mode)
- [x] Support to supply type codes
- [x] make format
- [x] Fix GLib binding
- [x] Fix Ruby binding
- [x] Fix Python binding
- [ ] Make CI green

### CODE REVIEW

**Comment by mrkn** [comment]
The reason of the CI failure is same as https://github.com/apache/arrow/pull/3760

**Comment by mrkn** [comment]
I think this pull-request is ready for review.

**Review by kou — COMMENTED** [review]
Can we create `arrow::UnionArray::Make()` instead of adding parameters for `union_()` to `MakeDense()` and `MakeSparse()`?

```cpp
static Status Make(const std::shared_ptr<DataType>& type,
                   const Array& type_ids,
                   const Array& value_offsets,
                   const std::vector<std::shared_ptr<Array>>& children,
                   std::shared_ptr<Array>* out);
```

I think that adding parameters for `union_()` to `MakeDense()` and `MakeSparse()` causes dirty API. If we accept `UnionType` directly, we can improve API by just improving `UnionType` API.

Other C++ developers, what do you think about this?

**Comment by mrkn** [comment]
@kou 

> Can we create `arrow::UnionArray::Make()` instead of adding parameters for `union_()` to `MakeDense()` and `MakeSparse()`?

My understanding of what you mean is:

1. the changes for MakeDense and MakeSparse should be reverted
2. we should add UnionType::Make so that we can create UnionArray by the code like below:

```cpp
arrow::UnionArray array;
auto array_type = arrow::union_(children, field_names, type_codes, UnionMode::SPARSE);
arrow::UnionArray::Make(array_type, type_ids, value_offsets, children, &array);
```

Is this interpretation correct?

**Comment by kou** [comment]
Yes. It's what I suggest.

**Comment by mrkn** [comment]
@wesm @pitrou @fsaintjacques How do you think about this PR?

**Comment by fsaintjacques** [comment]
I'll give a look.

**Comment by pitrou** [comment]
@mrkn It looks ok on the principle. Can you rebase on latest master?

**Comment by mrkn** [comment]
@pitrou I've rebased the branch onto the latest master, and fixed against two kou's comments.

I haven't added `arrow::UnionArray::Make` kou suggested, but I'd like to do it if the suggested approach is better than my approach.

**Comment by pitrou** [comment]
@mrkn Are you sure the rebase is ok? There seem to be some Rust conflicts...
Once the rebase is fixed, I'll take a look. Thank you :-)

**Comment by mrkn** [comment]
@pitrou Rebased.

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -408,8 +408,56 @@ def test_union_from_dense():
...
 
     result = pa.UnionArray.from_dense(types, value_offsets, [binary, int64])
+    ty = result.type
 
     assert result.to_pylist() == [b'a', 1, b'c', b'b', 2, 3, b'd']
+    assert [ty[i].name for i in range(ty.num_children)] == ['0', '1']
+    assert ty.type_codes == [0, 1]
+
+
+def test_union_from_dense_with_field_name():
```
Comment (pitrou) [inline python/pyarrow/tests/test_array.py]: Is it possible to consolidate all these tests? They don't differ by much...

File: python/pyarrow/array.pxi [hunk 2]
```diff
@@ -1084,10 +1098,19 @@ cdef class UnionArray(Array):
         cdef shared_ptr[CArray] out
         cdef vector[shared_ptr[CArray]] c
         cdef Array child
+        cdef vector[c_string] fn
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: Call this `c_field_names`?

File: python/pyarrow/array.pxi [hunk 3]
```diff
@@ -1084,10 +1098,19 @@ cdef class UnionArray(Array):
...
+        cdef vector[uint8_t] tc
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: Call this `c_type_codes`?

File: cpp/src/arrow/union_array-test.cc [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+using arrow::internal::checked_pointer_cast;
+
+namespace arrow {
+
+class TestUnionArray : public ::testing::Test {
+ public:
+  void SetUp() {
+    pool_ = default_memory_pool();
+    auto type_ids =
+        compute::_MakeArray<Int8Type, int8_t>(int8(), {0, 1, 2, 0, 1, 3, 2, 0, 2, 1}, {});
```
Comment (pitrou) [inline cpp/src/arrow/union_array-test.cc]: You should take a look at `ArrayFromJSON`. It generally makes writing such code easier.

File: cpp/src/arrow/union_array-test.cc [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <string>
+
+#include <gtest/gtest.h>
+
+#include "arrow/compute/test-util.h"
```
Comment (pitrou) [inline cpp/src/arrow/union_array-test.cc]: We shouldn't depend on `compute` utilities here.

File: cpp/src/arrow/union_array-test.cc [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+  std::shared_ptr<Array> result;
+  ASSERT_OK(UnionArray::MakeSparse(*type_ids_, children, &result));
+
+  UnionArray& union_array = checked_cast<UnionArray&>(*result);
+  ASSERT_EQ(UnionMode::SPARSE, union_array.mode());
+  CheckFieldNames(union_array, {"0", "1", "2", "3"});
+  CheckTypeCodes(union_array, {0, 1, 2, 3});
+}
+
+TEST_F(TestUnionArray, MakeDenseWithFieldNamesTest) {
```
Comment (pitrou) [inline cpp/src/arrow/union_array-test.cc]: Isn't there a way to condensate all these tests?

**Review by pitrou — CHANGES_REQUESTED** [review]
I can't comment on the glib or Ruby changes... Perhaps it's easier to make separate PRs for C++/Python and the rest?

**Comment by mrkn** [comment]
> Perhaps it's easier to make separate PRs for C++/Python and the rest?

@pitrou OK. I'll separate GLib and Ruby changes into a different pull request.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks @mrkn. The CI failures are unrelated (see ARROW-5148). I will make a few changes and then merge.

**Pull request closed without merging**